### PR TITLE
Add Ascend NPU support for nf4 quant

### PR DESCRIPTION
**Pull request by statelesshz**
### What does this PR do?
This PR adds Ascend NPU support for nf4 quant/dequant and allows QLoRA fine-tuning for LLMs using transformers, peft, and trl.

You may notice that the nf4 quantization method is currently implemented in PyTorch. This interim measure is due to the fact that the high-performance version implemented with AscendC is still in progress 😞 . Meanwhile, we've received feedback from many in the Ascend NPU community expressing their keen interest in using QLoRA to fine-tune LLMs as soon as possible, so there is this PR.

Related PR: https://github.com/huggingface/transformers/pull/31512

### Collaborators
@SlightwindSec @Ginray @MatrixPlayer

cc @Titus-von-Koeller @matthewdouglas

### CODE REVIEW

**Comment by statelesshz** [comment]
[asciicast](https://asciinema.org/a/wQ2SPSAbtAiAN7eqyHt5WnxPX)

**Comment by baymax591** [comment]
Thanks a lot for sharing this PR and the video demo! Thanks to the demo, I was able to successfully run NF4 quant/dequant on the NPU with ease. The detailed explanation in the video really helped me understand the process and key steps. Looking forward to more updates in the future—great work!

**Comment by baymax591** [comment]
I hope this PR can be merged soon, as it provides valuable improvements. Looking forward to seeing it merged!
cc @Titus-von-Koeller

**Comment by SunMarc** [comment]
Nice work and thanks for the demo ! Can you have a look @matthewdouglas ?

**Comment by matthewdouglas** [comment]
I will be able to look in more detail next week, but at first glance it looks nice. Thanks @statelesshz !

**Comment by matthewdouglas** [comment]
@statelesshz We really appreciate the contribution! Apart from a lint check, I think we can go ahead and merge this.

For awareness, we are currently planning to adopt usage of `torch.library` to register custom ops for multi-backend. When we've progressed further on that we will want to come back and migrate to an updated interface.

**Review by matthewdouglas — APPROVED** [review]

**Comment by statelesshz** [comment]
> @statelesshz We really appreciate the contribution! Apart from a lint check, I think we can go ahead and merge this.
> 
> For awareness, we are currently planning to adopt usage of `torch.library` to register custom ops for multi-backend. When we've progressed further on that we will want to come back and migrate to an updated interface.

@matthewdouglas Thank you for the feedback. I have addressed the lint check warnings, and I think the PR is now ready for merging. 🤗  
Could you please re-trigger the CI to ensure everything is in order?

File: bitsandbytes/autograd/_functions.py [hunk 1]
```diff
@@ -519,7 +519,12 @@ def forward(ctx, A, B, out=None, bias=None, quant_state: Optional[F.QuantState]
 
         # 1. Dequantize
         # 2. MatmulnN
-        output = torch.nn.functional.linear(A, F.dequantize_4bit(B, quant_state).to(A.dtype).t(), bias)
+        if A.device.type == "npu":
```
Comment (MengqingCao) [inline bitsandbytes/autograd/_functions.py:522]: A quick question: why not we use `torch.nn.functional.linear` directly? Thanks in advance for your answer

**Pull request merged**